compendium/DeclarativeServices/src/manager/ComponentFactoryImpl.cpp

```diff
@@ -0,0 +1,144 @@
+ /*=============================================================================
+
+  Library: CppMicroServices
+
+  Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+  file at the top-level directory of this distribution and at
+  https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  =============================================================================*/
+
+#include "../ComponentRegistry.hpp"
+#include "../metadata/ComponentMetadata.hpp"
+#include "../SCRExtensionRegistry.hpp"
+#include "ComponentConfigurationImpl.hpp"
+#include "ComponentFactoryImpl.hpp"
+#include "ComponentManagerImpl.hpp"
+#include "ConcurrencyUtil.hpp"
+#include "cppmicroservices/SecurityException.h"
+#include "cppmicroservices/SharedLibraryException.h"
+#include "cppmicroservices/asyncworkservice/AsyncWorkService.hpp"
+#include "cppmicroservices/cm/ConfigurationAdmin.hpp"
+
+
+namespace cppmicroservices
+{
+    namespace scrimpl
+    {
+        using cppmicroservices::scrimpl::metadata::ComponentMetadata;
+
+        ComponentFactoryImpl::ComponentFactoryImpl(
+            cppmicroservices::BundleContext const& context,
+            std::shared_ptr<cppmicroservices::logservice::LogService> logger,
+            std::shared_ptr<cppmicroservices::async::AsyncWorkService> asyncWorkSvc,
+            std::shared_ptr<SCRExtensionRegistry> extensionReg)
+            : bundleContext(context)
+            , logger(std::move(logger))
+            , asyncWorkService(asyncWorkSvc)
+            , extensionRegistry(extensionReg)
+        {
+            if (!bundleContext || !(this->logger) || (!this->asyncWorkService) || (!this->extensionRegistry))
+            {
+                throw std::invalid_argument("ComponentFactoryImpl Constructor "
+                                            "provided with invalid arguments");
+            }
+        }
+
+
+        void
+        ComponentFactoryImpl::CreateFactoryComponent(std::string const& pid,
+                                                     std::shared_ptr<ComponentConfigurationImpl>& mgr,
+                                                     cppmicroservices::AnyMap const& properties)
+        {
+            // Create the virtual metadata for the factory instance. 
+            // Start with the metadata from the factory 
+            auto oldMetadata = mgr->GetMetadata();
+            auto newMetadata = std::make_shared<ComponentMetadata>(*oldMetadata);
```

Comment:
maybe don't need the additional variable oldMetadata? instead could do:

`auto newMetadata = std::make_shared<ComponentMetadata>(*(mgr->GetMetadata()));` -- not sure what the 'correct' coding practice is here

Reply:
You are right. The extra variable is unecessary. Will fix.